www/src/components/users/MagicLogin.tsx

```diff
@@ -363,51 +222,156 @@ function OAuthOptions({ oauthUrls }: any) {
   const [email, setEmail] = useState('')
   const [password, setPassword] = useState('')
-  const [getLoginMethod, { data, loading: qLoading, error: qError }] = useLazyQuery(LOGIN_METHOD, {
-    variables: { email, host: host() },
-  })
+  const navigate = useNavigate()
+  const passwordRef = useRef<HTMLElement>()
+  const emailRef = useRef<HTMLElement>()
 
-  const loginMethod = data?.loginMethod?.loginMethod
-  const open = loginMethod === LoginMethod.PASSWORD
-  const passwordless = loginMethod === LoginMethod.PASSWORDLESS
+  useEffect(() => {
+    setInputFocus(emailRef)
+  }, [])
+
+  const [
+    loginMethodQuery,
+    {
+      data: loginMethodData,
+      loading: loginMethodLoading,
+      error: loginMethodError,
+    },
+  ] = useLoginMethodLazyQuery()
+  const getLoginMethod = useCallback(() => {
+    loginMethodQuery({
+      variables: { email, host: host() },
+    })
+  }, [email, loginMethodQuery])
+
+  const [loginMutation, { loading: loginMLoading, error: loginMError }]
+    = useLoginMutation({
+      variables: {
+        email,
+        password,
+        deviceToken: typeof deviceToken === 'string' ? deviceToken : undefined,
+      },
+      onCompleted: ({ login }) => {
+        setToken(login?.jwt)
+        if (deviceToken) finishedDeviceLogin()
+        if (challenge) {
+          handleOauthChallenge(client, challenge)
+        }
+        else {
+          history.navigate('/')
+        }
+      },
+    })
 
-  const { data: oAuthData } = useQuery(OAUTH_URLS, { variables: { host: host() } })
+  useEffect(() => {
+    if (state !== prevState.current) {
+      switch (state) {
+      case State.Initial:
+        setInputFocus(emailRef)
+        break
+      case State.CheckEmail:
+        getLoginMethod()
+        setState(State.CheckingEmail)
+        break
+      case State.PwdLogin_CheckPwd:
+        loginMutation()
+        setState(State.PwdLogin_CheckingPwd)
+        break
+      case State.PwdLogin:
+        setInputFocus(passwordRef)
+        break
+      default:
+        break
+      }
+    }
+    prevState.current = state
+  }, [getLoginMethod, loginMutation, state])
+
+  useEffect(() => {
+    if (state === State.Signup) {
+      navigate('/signup', { state: { email } })
+    }
+  })
+
+  useEffect(() => {
+    if (state === State.PwdLogin_CheckingPwd && loginMError) {
+      setState(State.PwdLogin)
+      setPassword('')
+    }
+  }, [loginMError, state])
+  const passwordErrorMsg
+    = loginMError?.message === 'invalid password' ? 'Invalid password' : undefined
+  const loginError = !passwordErrorMsg && loginMError
 
-  const [mutation, { loading: mLoading, error }] = useMutation(LOGIN_MUTATION, {
-    variables: { email, password, deviceToken },
-    onCompleted: ({ login: { jwt } }) => {
-      setToken(jwt)
-      if (deviceToken) finishedDeviceLogin()
-      if (challenge) {
-        handleOauthChallenge(client, challenge)
```

Comment:
do we still handle oauth here?

Reply:
That mutation just got moved up to line 285

Reply:
ahh